### ISSUE TITLE
Delete parts from C-sets

### ISSUE THREAD

**Issue by epatters** [AlgebraicJulia/Catlab.jl#196]
The last crucial feature missing from C-sets is deleting parts, via functions `delete_part!` and `delete_parts!`.

Deletion can be recursive or non-recursive (this terminology is probably nonstandard). For example, in the case of graphs, recursive deletion of a vertex would delete all edges incident to the vertex, whereas nonrecursive deletion would leave the source/target of those edges undefined. When the category C has infinitely many morphisms, as in the theory of discrete dynamical systems, recursive deletion of a single part can trigger the deletion of an arbitrarily long chain of additional parts.

To get started, it would be enough to implement non-recursive deletion and handle the case of graphs specially to conform to the LightGraphs convention.

**Comment by jpfairbanks** [comment]
Having the undefined behavior would require making the types able to handle `nothing` right? Could you have a `Vector{Int}` encoding the `src: E --> V` and then delete a vertex and now need to convert that to `Vector{Union{Int, Nothing}}`. That seems bad.

**Comment by epatters** [comment]
For implementation reasons I found it convenient to allow undefined parts (NULLs in the SQL analogy) and so I've actually allowed it from the beginning, although I didn't advertise it as a feature. Since the parts of a given type are always numbered from 1 to n, the NULLs can be represented simply by the integer 0. So we don't have to worry about `Vector{Int}` vs `Vector{Union{Int,Nothing}}`. The indexing logic for `set_subpart!` knows how to deal with these NULLs.
